File: ros2_workspaces/Books/ros2_from_scratch_by_Renard/part2_ws/src/my_py_pkg/my_py_pkg/count_until_server_minimal.py
#!/usr/bin/env python3
import rclpy
import time
from rclpy.node import Node
 # Recall that we can create subscriptions, servers and clients on self for Node. No so for ActionServer. We must import the following.
from rclpy.action import ActionServer, GoalResponse
from rclpy.action.server import ServerGoalHandle
from my_robot_interfaces.action import CountUntil


class CountUntilServerNode(Node):
    def __init__(self):
        super().__init__("count_util_server")
        # When creating the action server, we pass a reference to the node (self).
        self.count_util_server_ = ActionServer(
            self,
            CountUntil,
            "count_until",
            goal_callback=self.goal_callback,
            execute_callback=self.execute_callback)
        self.get_logger().info("Action server started!!")

    """
    Receive action request.
    Takes CountUntil.Goal request and returns GoalResponse ACCEPT or REJECT.
    ValidateS the target_number. 
    """
    def goal_callback(self, goal_request:CountUntil.Goal):
        self.get_logger().info("Received goal")
        if goal_request.target_number <= 0:
            self.get_logger().info("Rejecting the goal, target number must be positive")
            return GoalResponse.REJECT
        self.get_logger().info("Accepting the goal")
        return GoalResponse.ACCEPT
    
    # handle the goal
    def execute_callback(self, goal_handle: ServerGoalHandle):
        target_number = goal_handle.request.target_number
        delay = goal_handle.request.delay
        result = CountUntil.Result() # result to populate
        counter = 0

        self.get_logger().info("Executing the goal")
        for i in range(target_number):
            counter += 1
            self.get_logger().info(str(counter))
            time.sleep(delay)

        # Final state on the goal. During exeuction the state is executing.
        # Other possible final states are aborted and cancelled.
        goal_handle.succeed()
        result.reached_number = counter
        return result # return result part of interface

def main(args=None):
    rclpy.init(args=args)
    node = CountUntilServerNode() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_workspaces/Books/ros2_from_scratch_by_Renard/part2_ws/src/my_cpp_pkg/src/count_until_server_minimal.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/count_until.hpp"


using CountUntil = my_robot_interfaces::action::CountUntil;
using CountUntilGoalHandle = rclcpp_action::ServerGoalHandle<CountUntil>;
// contains special objects such as _1, _2, _3, etc. used with std::bind to represent arguments that will be provided later 
// when the resulting function object is called.
using namespace std::placeholders; 

class CountUntilServerNode : public rclcpp::Node {

    public:
    CountUntilServerNode() : Node("count_until_server_node_cpp") {
        count_until_server_ = rclcpp_action::create_server<CountUntil>(
            this,
            "count_until",
            std::bind(&CountUntilServerNode::goalCallback, this,_1,_2),
            std::bind(&CountUntilServerNode::cancelCallback, this,_1),
            std::bind(&CountUntilServerNode::executeCallback, this,_1)
        );
        RCLCPP_INFO(this->get_logger(), "Action server started!!");
    }


    private:
    rclcpp_action::GoalResponse goalCallback(const rclcpp_action::GoalUUID &uuid, std::shared_ptr<const CountUntil::Goal> goal) {
        (void)uuid;
        RCLCPP_INFO(this->get_logger(), "Received goal");
        if (goal->target_number <= 0) {
            RCLCPP_WARN(this->get_logger(),"Rejecting goal, target number must be positive");
            return rclcpp_action::GoalResponse::REJECT;
        }
        RCLCPP_INFO(this->get_logger(),"Accepting the goal");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse cancelCallback(const std::shared_ptr<CountUntilGoalHandle> goal_handle) {
        (void)goal_handle;
        RCLCPP_INFO(this->get_logger(), "Received a cancel request");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void executeCallback(const std::shared_ptr<CountUntilGoalHandle> goal_handle) {
        int target_number = goal_handle->get_goal()->target_number;
        double delay = goal_handle->get_goal()->delay;
        auto result = std::make_shared<CountUntil::Result>();
        int counter = 0;
        rclcpp::Rate loop_rate(1.0/delay);

        RCLCPP_INFO(this->get_logger(), "Executing the goal");
        for (int i = 0; i < target_number; i++) {
            counter++;
            RCLCPP_INFO(this->get_logger(), "%d", counter);
            loop_rate.sleep();
        }
        result->reached_number = counter;
        goal_handle->succeed(result);
    }

    rclcpp_action::Server<CountUntil>::SharedPtr count_until_server_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilServerNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}